File: 3. Behavioral Design Patterns/Template Method/Go/main.go
package main

import "fmt"

// Template Method Pattern Example in Go
// DataProcessor with CSV and JSON implementations

type DataProcessor interface {
	ProcessData()
	LoadData()
	ProcessDataImpl()
	SaveData()
}

type BaseProcessor struct {
	impl DataProcessor
}

func (b *BaseProcessor) ProcessData() {
	b.impl.LoadData()
	b.impl.ProcessDataImpl()
	b.impl.SaveData()
}

type CSVProcessor struct {
	BaseProcessor
}

func NewCSVProcessor() *CSVProcessor {
	p := &CSVProcessor{}
	p.impl = p
	return p
}

func (c *CSVProcessor) LoadData()        { fmt.Println("Loading CSV data") }
func (c *CSVProcessor) ProcessDataImpl() { fmt.Println("Processing CSV data") }
func (c *CSVProcessor) SaveData()        { fmt.Println("Saving CSV data") }

type JSONProcessor struct {
	BaseProcessor
}

func NewJSONProcessor() *JSONProcessor {
	p := &JSONProcessor{}
	p.impl = p
	return p
}

func (j *JSONProcessor) LoadData()        { fmt.Println("Loading JSON data") }
func (j *JSONProcessor) ProcessDataImpl() { fmt.Println("Processing JSON data") }
func (j *JSONProcessor) SaveData()        { fmt.Println("Saving JSON data") }

func main() {
	csv := NewCSVProcessor()
	csv.ProcessData()
	fmt.Println()
	json := NewJSONProcessor()
	json.ProcessData()
}

File: 3. Behavioral Design Patterns/Template Method/Rust/main.rs
// Template Method Pattern Example in Rust
// DataProcessor with CSV and JSON implementations

trait DataProcessor {
    fn process_data(&self) {
        self.load_data();
        self.process_data_impl();
        self.save_data();
    }
    fn load_data(&self);
    fn process_data_impl(&self);
    fn save_data(&self);
}

struct CSVProcessor;
impl DataProcessor for CSVProcessor {
    fn load_data(&self) { println!("Loading CSV data"); }
    fn process_data_impl(&self) { println!("Processing CSV data"); }
    fn save_data(&self) { println!("Saving CSV data"); }
}

struct JSONProcessor;
impl DataProcessor for JSONProcessor {
    fn load_data(&self) { println!("Loading JSON data"); }
    fn process_data_impl(&self) { println!("Processing JSON data"); }
    fn save_data(&self) { println!("Saving JSON data"); }
}

fn main() {
    let csv = CSVProcessor;
    csv.process_data();
    println!();
    let json = JSONProcessor;
    json.process_data();
}
